lucene/core/src/java/org/apache/lucene/store/Directory.java

```diff
@@ -48,6 +49,12 @@
  * @see FilterDirectory
  */
 public abstract class Directory implements Closeable {
+
+  /**
+   * Reader only attributes for this directory.
+   */
+  private Map<String, String> readerAttributes;
```

Comment:
I don't think `Directory` is the right abstraction to hold this. I suggest we add an overload to DirectoryReader.open that looks like this:

```Java
  public static DirectoryReader open(final Directory directory, final Map<String, String> attributes) throws IOException;

//
  public static DirectoryReader open(final IndexCommit commit, final Map<String, String> attributes) throws IOException;
```
that allows us to have a dedicated map of attributes per open call. Subsequent reopens then inherit the attributes form the reader. You can also then make this a property of `DirecotryReader` I guess.

Reply:
>I don't think Directory is the right abstraction to hold this. 

Is that because single directory can be opened in multiple contexts?

>Subsequent reopens then inherit the attributes form the reader. You can also then make this a property of DirecotryReader I guess.

How is the attribute map passed to SegmentReadState in this approach? Also, IndexWriterConfig need not be changed with this approach, right?

Reply:
> Is that because single directory can be opened in multiple contexts?
I think there are multiple reasons I think it should be either bound to a single reader or writer. ie two calls to DirectoryReader.open can return readers with different fields on-heap even if they share the same directory.

> How is the attribute map passed to SegmentReadState in this approach? Also, IndexWriterConfig need not be changed with this approach, right?

I don't think IWC needs to change. I think such a map can be passed down to `SegmentCoreReaders` and then passed on via `SegmentReadState` I am not sure why you think we need to change a lot of things?

Reply:
>I think such a map can be passed down to SegmentCoreReaders and then passed on via SegmentReadState I am not sure why you think we need to change a lot of things?

Maybe, better way is to add `readerAttributes` in `SegmentInfo`, and populate that during `StandardDirectoryReader.open`?